certification/pyxis/types.go

```diff
@@ -1,10 +1,201 @@
+
+	err = b.storePreflightResults(preflightResultsFilepath)
+	if err != nil {
+		return nil, err
+	}
+
+	return &b, nil
+}
+
+// storeRPMManifest reads the manifest from disk at path and stores it in
+// the CertificationInput as an RPMManifest struct.
+func (b *certificationInputBuilder) storeRPMManifest(filepath string) error {
+	bytes, err := os.ReadFile(filepath)
+	if err != nil {
+		return fmt.Errorf(
+			"%w: unable to read file from disk to include in submission: %s: %s",
+			errors.ErrSubmittingToPyxis,
+			filepath,
+			err,
+		)
+	}
+
+	var manifest RPMManifest
+	err = json.Unmarshal(bytes, &manifest)
+	if err != nil {
+		return fmt.Errorf(
+			"%w: data for the %s appears to be malformed: %s",
+			errors.ErrSubmittingToPyxis,
+			"rpm manifest",
+			err,
+		)
+	}
+
+	b.RpmManifest = &manifest
+	return nil
+}
+
+// storePreflightResults reads the results from disk at path and stores it in
+// the CertificationInput as TestResults.
+func (b *certificationInputBuilder) storePreflightResults(filepath string) error {
+	bytes, err := os.ReadFile(filepath)
+	if err != nil {
+		return fmt.Errorf(
+			"%w: unable to read file from disk to include in submission: %s: %s",
+			errors.ErrSubmittingToPyxis,
+			filepath,
+			err,
+		)
+	}
+
+	var testResults TestResults
+	err = json.Unmarshal(bytes, &testResults)
+	if err != nil {
+		return fmt.Errorf(
+			"%w: data for the %s appears to be malformed: %s",
+			errors.ErrSubmittingToPyxis,
+			"preflight results.json",
+			err,
+		)
+	}
+
+	b.TestResults = &testResults
+	return nil
+}
+
+// storeCertImage reads the image from disk at path and stores it in
+// the CertificationInput as a CertImage
+func (b *certificationInputBuilder) storeCertImage(filepath string) error {
+	bytes, err := os.ReadFile(filepath)
+	if err != nil {
+		return fmt.Errorf(
+			"%w: unable to read file from disk to include in submission: %s: %s",
+			errors.ErrSubmittingToPyxis,
+			filepath,
+			err,
+		)
+	}
+
+	var image CertImage
+	err = json.Unmarshal(bytes, &image)
+	if err != nil {
+		return fmt.Errorf(
+			"%w: data for the %s appears to be malformed: %s",
+			errors.ErrSubmittingToPyxis,
+			"certImage",
+			err,
+		)
+	}
+
+	b.CertImage = &image
+	return nil
+}
+
+// Finalize runs a collection of safeguards to try to ensure we get a reliable
+// CertificationInput. This also wires up information that's shared across
+// the various included assets (e.g. ISVPID) where applicable, and returns an
+// unmodifiable CertificationInput
+func (b *certificationInputBuilder) Finalize() (*certificationInput, error) {
+	// safeguards, make sure things aren't nil for any reason.
+	if b.CertImage == nil {
+		return nil, fmt.Errorf("a CertImage was not provided and is required")
+	}
+	if b.TestResults == nil {
+		return nil, fmt.Errorf("test results were not provided and are required")
+	}
+
+	if b.RpmManifest == nil {
+		return nil, fmt.Errorf("the RPM manifest was not provided and is required")
+	}
+
+	if b.Artifacts == nil {
```

Comment:
I think for now since we are required to send the logs, maybe we make sure the length is one? Then over time increment this as we add more things? Maybe that makes less sense than what is here since we'd have to come back and remember about this finalizer.

Reply:
Yeah I agree, just trying to think of better ways to make sure we send what we should. From preflights prospective they are optional, but not from connects standpoint. They expect to get artifacts all the time.

Reply:
I don't think that belongs in preflight.